executor/new_index_lookup_join.go

```diff
@@ -0,0 +1,531 @@
+// It preserves the order of the outer table and support batch lookup.
+//
+// The execution flow is very similar to IndexLookUpReader:
+// 1. outerWorker read N outer rows, build a task and send it to result channel and inner worker channel.
+// 2. The innerWorker receives the task, builds key ranges from outer rows and fetch inner rows, builds inner row hash map.
+// 3. main thread receives the task, waits for inner worker finish handling the task.
+// 4，main thread join each outer row by look up the inner rows hash map in the task.
+type NewIndexLookUpJoin struct {
+	baseExecutor
+
+	resultCh   <-chan *lookUpJoinTask
+	cancelFunc goctx.CancelFunc
+	workerWg   *sync.WaitGroup
+
+	outerCtx outerCtx
+	innerCtx innerCtx
+
+	task             *lookUpJoinTask
+	joinResult       *chunk.Chunk
+	joinResultCursor int
+
+	resultGenerator joinResultGenerator
+}
+
+type outerCtx struct {
+	schema   *expression.Schema
+	rowTypes []*types.FieldType
+	keyCols  []int
+	filter   expression.CNFExprs
+}
+
+type innerCtx struct {
+	readerBuilder *dataReaderBuilder
+	rowTypes      []*types.FieldType
+	keyCols       []int
+	filter        expression.CNFExprs
+}
+
+type lookUpJoinTask struct {
+	outerResult *chunk.Chunk
+	outerMatch  []bool
+
+	innerResult       *chunk.List
+	encodedLookUpKeys *chunk.Chunk
+	lookupMap         *mvmap.MVMap
+	matchedInners     []chunk.Row
+
+	doneCh chan error
+	cursor int
+}
+
+type outerWorker struct {
+	outerCtx
+
+	ctx      context.Context
+	executor Executor
+
+	executorChk *chunk.Chunk
+
+	maxBatchSize int
+	batchSize    int
+
+	resultCh chan<- *lookUpJoinTask
+	innerCh  chan<- *lookUpJoinTask
+}
+
+type innerWorker struct {
+	innerCtx
+
+	taskCh      <-chan *lookUpJoinTask
+	outerCtx    outerCtx
+	ctx         context.Context
+	executorChk *chunk.Chunk
+}
+
+// Open implements the Executor interface.
+func (e *NewIndexLookUpJoin) Open(goCtx goctx.Context) error {
+	err := e.children[0].Open(goCtx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.startWorkers(goCtx)
+	return nil
+}
+
+func (e *NewIndexLookUpJoin) startWorkers(goCtx goctx.Context) {
+	concurrency := e.ctx.GetSessionVars().IndexLookupConcurrency
+	resultCh := make(chan *lookUpJoinTask, concurrency)
+	e.resultCh = resultCh
+	workerCtx, cancelFunc := goctx.WithCancel(goCtx)
+	e.cancelFunc = cancelFunc
+	innerCh := make(chan *lookUpJoinTask, concurrency)
+	e.workerWg.Add(1)
+	go e.newOuterWorker(resultCh, innerCh).run(workerCtx, e.workerWg)
+	e.workerWg.Add(concurrency)
+	for i := 0; i < concurrency; i++ {
+		go e.newInnerWorker(innerCh).run(workerCtx, e.workerWg)
+	}
+}
+
+func (e *NewIndexLookUpJoin) newOuterWorker(resultCh, innerCh chan *lookUpJoinTask) *outerWorker {
+	ow := &outerWorker{
+		outerCtx:     e.outerCtx,
+		ctx:          e.ctx,
+		executor:     e.children[0],
+		executorChk:  chunk.NewChunk(e.outerCtx.rowTypes),
+		resultCh:     resultCh,
+		innerCh:      innerCh,
+		batchSize:    32,
+		maxBatchSize: e.ctx.GetSessionVars().IndexLookupSize,
```

Comment:
`IndexLookupSize` -> `IndexJoinBatchSize`, the first one is for the second read of double read executor.